apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/RpcRetryConsumerTest.cs

```diff
@@ -0,0 +1,223 @@
+                retryConsumer.Receive(intArray);
+
+                timer.Call();
+                timer.Call();
+                // Extra call to ensure buffer is emptied.
+                timer.Call();
+                mockConsumer.Verify(c => c.Receive(intArray), Times.Exactly(3));
+            }, RetryOptions.Retry(retryAttempts: 2));
+        }
+
+        [Fact]
+        public void Receive_Retry_MaxRetries_Throw()
+        {
+            SetupTestReceiveThrows((timer, mockConsumer, retryConsumer) =>
+            {
+                int[] intArray = { 1, 2, 3, 4 };
+                retryConsumer.Receive(intArray);
+
+                timer.Call();
+                Assert.Throws<InvalidOperationException>(() => { timer.Call(); });
+                // Extra call to ensure buffer is emptied.
+                timer.Call();
+                mockConsumer.Verify(c => c.Receive(intArray), Times.Exactly(3));
+            }, RetryOptions.Retry(exceptionHandling: ExceptionHandling.Propagate, retryAttempts: 2));
+        }
+
+        /// <summary>
+        /// Creates a timer factory with a fake timer.
+        /// </summary>
+        private Func<Action, RetryOptions, ISequentialThreadingTimer> CreateTimerFactory(FakeSequentialThreadingTimer timer)
+        {
+            return (callback, retryOptions) =>
+            {
+                timer.Initialize(callback, retryOptions.RetryInterval);
+                return timer;
+            };
+        }
+
+        /// <summary>
+        /// Sets up a test where all calls to the consumer passed to the <see cref="RpcRetryConsumer"/> will throw
+        /// an <see cref="RpcException"/>.
+        /// </summary>
+        private void SetupTestReceiveThrows(Action<FakeSequentialThreadingTimer, Mock<IConsumer<int>>, RpcRetryConsumer<int>> action, RetryOptions options)
+        {
+            var timer = new FakeSequentialThreadingTimer();
+            var mockConsumer = new Mock<IConsumer<int>>();
+            var timerFactory = CreateTimerFactory(timer);
+            mockConsumer.Setup(c => c.Receive(It.IsAny<IEnumerable<int>>()))
+                    .Throws(new RpcException(Status.DefaultSuccess));
+            using (var retryConsumer = new RpcRetryConsumer<int>(
+                mockConsumer.Object, options, Utils.ConstantSizer<int>.GetSize, timerFactory))
+            {
+                action(timer, mockConsumer, retryConsumer);
+            }
+        }
+
+        /// <summary>
+        /// A fake <see cref="ISequentialThreadingTimer"/> where the callback is only called when
+        /// <see cref="Call"/> is called.
+        /// </summary>
+        private class FakeSequentialThreadingTimer : ISequentialThreadingTimer
+        {
+            private Action _callback;
+
+            public void Initialize(Action callback, TimeSpan waitTime) => _callback = callback;
+
+            public void Call() => GaxPreconditions.CheckNotNull(_callback, nameof(_callback))();
```

Comment:
I feel like this check isn't really necessary. Are you trying to guard against forgetting to initialize the timer? You'll find that out pretty quickly when running the tests.